src/cockpit/unixtransport.py

```diff
@@ -0,0 +1,56 @@
+# This file is part of Cockpit.
+#
+# Copyright (C) 2022 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import os
+
+
+class UnixTransport:
+    BLOCK_SIZE = 1024 * 1024
+
+    def __init__(self, loop, in_fd, out_fd, closefds=False):
+        self.loop = loop
+        self.in_fd = in_fd
+        self.out_fd = out_fd
+        self.to_close = set()
+        if closefds:
+            self.to_close.update((in_fd, out_fd))
+        self.connection_lost = loop.create_future()
+
+    def read(self):
+        data = os.read(self.in_fd, UnixTransport.BLOCK_SIZE)
+        self.protocol.data_received(data)
+        if data == b'':
+            self.protocol.eof_received()
+            self.close()
+
+    def write(self, data):
+        os.write(self.out_fd, data)
```

Comment:
Since py 3.5 this is now [safe against EINTR](https://docs.python.org/3/library/os.html#os.write), but we could still have short writes, no? Doesn't that still need a retry loop? Alternatively, could we go one level up and read from/write to `sys.std{in,out}.buffer`?

Reply:
I ended up going "full strength" and implementing a write queue, complete with backpressure, polling for writability, and draining the queue on connection close.